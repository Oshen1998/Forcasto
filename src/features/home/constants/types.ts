export type IWeather = {
  id: 501;
  main: string;
  description: string;
  icon: string;
};

export interface IWeatherInfo {
  weather: IWeather[];
  name: string; // location name
  sys: {
    country: string;
  };
  main: {
    temp: number;
    pressure: number;
    humidity: number;
  };
  wind: {
    speed: number;
    deg: number;
    gust: number;
  };
  clouds: {
    all: number;
  };
}
