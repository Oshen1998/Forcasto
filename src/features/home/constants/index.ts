export const ICON_CODES = {
  CLEAR_DAY: '01d',
  CLEAR_NIGHT: '01n',

  RAIN_NIGHT: '09n',
  RAIN_DAY: '09d',

  _RAIN_NIGHT: '10n',
  _RAIN_DAY: '10d',

  SNOW_DAY: '13d',
  SNOW_NIGHT: '13n',
};
