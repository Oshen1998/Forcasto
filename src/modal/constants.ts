export enum MODAL_STACK {
  NOTIFY = 'NotifyModal',
  LOADER = 'LoaderModal',
}
