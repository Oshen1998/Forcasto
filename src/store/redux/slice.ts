import {createAction, createSlice} from '@reduxjs/toolkit';
import {AppActionTypes} from './types';

export interface AppState {
  count: number;
}

const initialState: AppState = {
  count: 0,
};

const AppSlice = createSlice({
  name: 'App',
  initialState,
  reducers: {
    countIncrement: state => {
      state.count = state.count + 1;
    },
    countDecrement: state => {
      state.count = state.count - 1;
    },
  },
});

const appSagaActions = {
  requestCurrentWeatherData: createAction(
    AppActionTypes.REQUEST_CURRENT_WEATHER,
  ),
};

export const AppActions = {
  ...AppSlice.actions,
  ...appSagaActions,
};
export default AppSlice.reducer;
