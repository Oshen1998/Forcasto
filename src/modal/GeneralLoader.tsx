import React from 'react';
import {StyleSheet, View} from 'react-native';
import {ModalComponentProp} from 'react-native-modalfy';
import {MODAL_STACK} from './constants';
import {ModalStackParams} from './modal.types';
import {deviceHeight, deviceWidth} from '../utils/device.utils';
import {LightColors} from '../themes/colors';
import {FONT_FAMILIES, FONT_SIZES} from '../themes/fonts';
import AppLottie from '../components/lottier/AppLottie';
import {animations} from '../themes/animations';
import AppText from '../components/text/AppText';

type LoaderModalProps = ModalComponentProp<
  ModalStackParams,
  void,
  MODAL_STACK.LOADER
>;

const GeneralLoader = ({modal}: LoaderModalProps) => {
  const {params} = modal;

  return (
    <View
      style={[
        styles.container,
        {backgroundColor: LightColors.Background.PRIMARY},
      ]}
      renderToHardwareTextureAndroid>
      <View style={styles.animationContainer}>
        <AppLottie source={animations.cloudLoading} height={150} width={150} />
      </View>
      <AppText
        testId="general-loader-id"
        textAlign="center"
        style={styles.desc}
        numberOfLines={2}>
        {params?.payload?.description ?? ''}
      </AppText>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    borderTopRightRadius: 20,
    borderTopLeftRadius: 20,
    marginHorizontal: 15,
    overflow: 'hidden',
    width: deviceWidth,
    maxHeight: deviceHeight * 0.6,
    alignItems: 'center',
    paddingHorizontal: 24,
    paddingVertical: 20,
  },
  title: {
    fontFamily: FONT_FAMILIES.Manrope.SemiBold,
    fontSize: FONT_SIZES.Subtitle,
    textAlign: 'center',
  },
  animation: {
    width: '100%',
    height: '100%',
  },
  animationContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 150,
    width: '100%',
  },
  desc: {
    marginVertical: 10,
    fontFamily: FONT_FAMILIES.Manrope.Medium,
    fontSize: FONT_SIZES.Caption,
    textAlign: 'center',
    color: LightColors.Text.DESCRIPTION,
  },
});

export default GeneralLoader;
