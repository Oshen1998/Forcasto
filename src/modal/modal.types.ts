import {MODAL_STACK} from './constants';

export interface ILoader {
  title: string;
  description?: string;
}

export interface ModalStackParams {
  [MODAL_STACK.LOADER]: {
    payload: ILoader;
  };
}
