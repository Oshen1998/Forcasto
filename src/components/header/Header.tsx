import {
  ColorValue,
  Image,
  StatusBar,
  StyleSheet,
  TouchableOpacity,
  View,
} from 'react-native';
import React, {memo, useCallback} from 'react';
import {useAppDispatch, useAppSelector} from '../../hooks/useRedux';
import {currentWeatherInfoSelector} from '../../store/redux/selector';
import AppText from '../text/AppText';
import {LightColors} from '../../themes/colors';
import {useSafeAreaInsets} from 'react-native-safe-area-context';
import {images} from '../../themes/images';
import {FONT_FAMILIES, FONT_SIZES} from '../../themes/fonts';
import {AppActions} from '../../store/redux/slice';
import {openBottomSheet} from '../../utils/modal.utils';
import {INotify} from '../../modal/modal.types';
import {MODAL_STACK} from '../../modal/constants';

export type HeaderProps = {
  colors: {
    backgroundColor: string | ColorValue;
    fontColor: string | ColorValue;
  };
};

const Header = ({colors}: HeaderProps) => {
  const dispatch = useAppDispatch();
  const insets = useSafeAreaInsets();
  const {name, sys} = useAppSelector(currentWeatherInfoSelector);

  // Hope to develop map integration part
  // user can select the location through the map
  const onHandlePressLocation = openBottomSheet<{payload: INotify}>(
    MODAL_STACK.NOTIFY,
    {
      payload: {
        title: 'Coming Soon!',
        animationKey: 'MAINTENANCE',
        description:
          "Thank you for your interest! Our app is in development, and we're excited to bring you an amazing experience soon. Stay tuned for the official release!",
      },
    },
  );

  const onHandlePressReload = useCallback(() => {
    dispatch(AppActions.reloadCurrentWeatherData());
  }, [dispatch]);

  return (
    <View style={[styles.container, {paddingTop: insets.top + 20}]}>
      <StatusBar translucent backgroundColor={LightColors.TRANSPARENT} />
      <TouchableOpacity
        onPress={onHandlePressLocation}
        style={styles.searchLocation}>
        <Image
          source={images.icons.location}
          resizeMode="contain"
          style={[styles.location, {tintColor: colors.fontColor}]}
        />
        <AppText
          textAlign="left"
          style={[styles.headerText, {color: colors.fontColor}]}>
          {name} ({sys.country})
        </AppText>
        <Image
          source={images.icons.chevron}
          resizeMode="contain"
          style={[styles.chevron, {tintColor: colors.fontColor}]}
        />
      </TouchableOpacity>
      <TouchableOpacity
        onPress={onHandlePressReload}
        style={styles.searchLocation}>
        <Image
          source={images.icons.reload}
          resizeMode="contain"
          style={[styles.reload, {tintColor: colors.fontColor}]}
        />
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: '100%',
  },
  headerText: {
    fontSize: FONT_SIZES.MediumTitle,
    fontFamily: FONT_FAMILIES.Manrope.SemiBold,
  },
  searchLocation: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 10,
  },
  location: {
    height: 20,
    width: 20,
    marginHorizontal: 10,
  },
  chevron: {
    height: 35,
    width: 35,
    marginTop: 5,
  },
  reload: {
    height: 20,
    width: 20,
    marginHorizontal: 10,
  },
});

export default memo(Header);
