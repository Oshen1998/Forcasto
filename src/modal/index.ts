import {
  createModalStack,
  ModalOptions,
  ModalStackConfig,
} from 'react-native-modalfy';
import {slideFromBottomAnimation} from './modal.utils';
import {MODAL_STACK} from './constants';
import GeneralLoader from './GeneralLoader';

const modalConfig: ModalStackConfig = {
  [MODAL_STACK.LOADER]: {
    modal: GeneralLoader,
    backBehavior: 'pop',
  },
};

const defaultOptions: ModalOptions = {
  disableFlingGesture: true,
  position: 'bottom',
  transitionOptions: slideFromBottomAnimation,
};

export const ModalStack = createModalStack(modalConfig, defaultOptions);
