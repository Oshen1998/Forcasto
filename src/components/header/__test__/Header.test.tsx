import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react-native';
import Header from '../Header';
import {openBottomSheet} from '../../../utils/modal.utils';
import {AppActions} from '../../../store/redux/slice';
import {useAppDispatch, useAppSelector} from '../../../hooks/useRedux';
import {MODAL_STACK} from '../../../modal/constants';

// Those mocks should be inside of mock file (much better)
jest.mock('../../../hooks/useRedux.ts');

jest.mock('react-native-safe-area-context', () => ({
  useSafeAreaInsets: jest.fn().mockReturnValue({top: 0, bottom: 0}),
}));

jest.mock('../../../utils/modal.utils.ts', () => ({
  openBottomSheet: jest.fn(),
}));

jest.mock('../../../store/redux/selector', () => ({
  currentWeatherInfoSelector: jest.fn().mockReturnValue({
    name: 'Colombo',
    sys: {country: 'LK'},
  }),
  appColorSelector: jest.fn().mockReturnValue({
    fontColor: '#000000',
    backgroundColor: '#C5E8F5',
  }),
}));

jest.mock('../../../themes/images', () => ({
  images: {
    icons: {
      location: 'mock-location-icon',
      chevron: 'mock-chevron-icon',
      reload: 'mock-reload-icon',
    },
  },
}));

describe('Header Component', () => {
  const dispatchMock = jest.fn();
  beforeEach(() => {
    (useAppDispatch as unknown as jest.Mock).mockReturnValue(dispatchMock);
    (useAppSelector as unknown as jest.Mock).mockReturnValue({
      name: 'Colombo',
      sys: {country: 'LK'},
    });
    jest.clearAllMocks();
  });

  it('should renders correctly', () => {
    render(<Header colors={{backgroundColor: '#000', fontColor: '#0ff'}} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('should renders location and reload buttons correctly', () => {
    const {getByText, getByTestId} = render(
      <Header colors={{backgroundColor: '#000', fontColor: '#0ff'}} />,
    );

    // Check if location button displays city name and country
    expect(getByText('Colombo (LK)')).toBeTruthy();

    // Check if the reload button exists
    expect(getByTestId('reload-button')).toBeTruthy();
  });

  it('should open the bottom sheet when location is pressed', () => {
    const {getByTestId} = render(
      <Header colors={{backgroundColor: '#000', fontColor: '#0ff'}} />,
    );

    const locationButton = getByTestId('location-button');
    fireEvent.press(locationButton);

    expect(openBottomSheet).toHaveBeenCalledWith(MODAL_STACK.NOTIFY, {
      payload: {
        title: 'Coming Soon!',
        animationKey: 'MAINTENANCE',
        description:
          "Thank you for your interest! Our app is in development, and we're excited to bring you an amazing experience soon. Stay tuned for the official release!",
      },
    });
  });

  it('should dispatch the reload action when reload button is pressed', () => {
    const {getByTestId} = render(
      <Header colors={{backgroundColor: '#000', fontColor: '#0ff'}} />,
    );

    const reloadButton = getByTestId('reload-button');
    fireEvent.press(reloadButton);

    expect(dispatchMock).toHaveBeenCalledWith(
      AppActions.reloadCurrentWeatherData(),
    );
  });
});
