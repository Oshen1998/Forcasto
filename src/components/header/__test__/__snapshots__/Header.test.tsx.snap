// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component renders correctly 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      },
      {
        "paddingTop": 20,
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 10,
        "opacity": 1,
      }
    }
    testID="location-button"
  >
    <Image
      resizeMode="contain"
      source="mock-location-icon"
      style={
        [
          {
            "height": 20,
            "marginHorizontal": 10,
            "width": 20,
          },
          {
            "tintColor": "#0ff",
          },
        ]
      }
    />
    <Text
      accessibilityLabel="header-country"
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        [
          {
            "fontFamily": "Manrope-SemiBold",
            "fontSize": 22,
          },
          {
            "color": "#0ff",
          },
        ]
      }
      testID="header-country"
    >
      Colombo
       (
      LK
      )
    </Text>
    <Image
      resizeMode="contain"
      source="mock-chevron-icon"
      style={
        [
          {
            "height": 35,
            "marginTop": 5,
            "width": 35,
          },
          {
            "tintColor": "#0ff",
          },
        ]
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 10,
        "opacity": 1,
      }
    }
    testID="reload-button"
  >
    <Image
      resizeMode="contain"
      source="mock-reload-icon"
      style={
        [
          {
            "height": 20,
            "marginHorizontal": 10,
            "width": 20,
          },
          {
            "tintColor": "#0ff",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Header Component should renders correctly 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      },
      {
        "paddingTop": 20,
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 10,
        "opacity": 1,
      }
    }
    testID="location-button"
  >
    <Image
      resizeMode="contain"
      source="mock-location-icon"
      style={
        [
          {
            "height": 20,
            "marginHorizontal": 10,
            "width": 20,
          },
          {
            "tintColor": "#0ff",
          },
        ]
      }
    />
    <Text
      accessibilityLabel="header-country"
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        [
          {
            "fontFamily": "Manrope-SemiBold",
            "fontSize": 22,
          },
          {
            "color": "#0ff",
          },
        ]
      }
      testID="header-country"
    >
      Colombo
       (
      LK
      )
    </Text>
    <Image
      resizeMode="contain"
      source="mock-chevron-icon"
      style={
        [
          {
            "height": 35,
            "marginTop": 5,
            "width": 35,
          },
          {
            "tintColor": "#0ff",
          },
        ]
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 10,
        "opacity": 1,
      }
    }
    testID="reload-button"
  >
    <Image
      resizeMode="contain"
      source="mock-reload-icon"
      style={
        [
          {
            "height": 20,
            "marginHorizontal": 10,
            "width": 20,
          },
          {
            "tintColor": "#0ff",
          },
        ]
      }
    />
  </View>
</View>
`;
